baseplate/frameworks/queue_consumer/kafka.py

```diff
@@ -0,0 +1,448 @@
+                    return
+
+                try:
+                    ingest_timestamp_ms = data["endpoint_timestamp"]
+                    now_ms = int(time.time() * 1000)
+                    message_latency = (now_ms - ingest_timestamp_ms) / 1000
+                except KeyError:
+                    # we can't guarantee that all publishers populate this field
+                    # v2 events publishers (event collectors) do, but future
+                    # kafka publishers may not
+                    message_latency = None
+
+                self.handler_fn(context, data, message)
+
+                if self.on_success_fn:
+                    self.on_success_fn(context, data, message)
+
+                if message_latency is not None:
+                    context.metrics.timer(f"{self.name}.{topic}.latency").send(message_latency)
+
+                context.metrics.gauge(f"{self.name}.{topic}.offset.{partition}").replace(offset)
+        except Exception:
+            # let this exception crash the server so we'll stop processing messages
+            # and won't commit offsets. when the server restarts it will get
+            # this message again and try to process it.
+            logger.exception(
+                "Unhandled error while trying to process a message, terminating the server"
+            )
+            raise
+
+
+class _BaseKafkaQueueConsumerFactory(QueueConsumerFactory):
+    def __init__(
+        self,
+        name: str,
+        baseplate: Baseplate,
+        bootstrap_servers: str,
+        group_id: str,
+        topics: Sequence[str],
+        handler_fn: Handler,
+        message_unpack_fn: Callable[[bytes], Dict[str, Any]] = json.loads,
+        health_check_fn: Optional[HealthcheckCallback] = None,
+    ):
+        self.baseplate = baseplate
+
+        service_name, _, group_name = group_id.partition(".")
+        assert service_name and group_name, "group_id must start with 'SERVICENAME.'"
+        assert name == f"kafka_consumer.{group_name}"
+
+        self.consumer = self._make_consumer(bootstrap_servers, group_id, topics)
+        self.name = f"kafka_consumer.{group_name}"
+
+        self.handler_fn = handler_fn
+        self.message_unpack_fn = message_unpack_fn
+        self.health_check_fn = health_check_fn
+
+    @classmethod
+    def _consumer_config(cls) -> Dict[str, Any]:
+        raise NotImplementedError
+
+    @classmethod
+    def _make_consumer(
+        cls, bootstrap_servers: str, group_id: str, topics: Sequence[str]
+    ) -> confluent_kafka.Consumer:
+        consumer_config = {
+            "bootstrap.servers": bootstrap_servers,
+            "group.id": group_id,
+            # reset the offset to the latest offset when no stored offset exists.
+            # this means that when a new consumer group is created it will only
+            # process new messages.
+            "auto.offset.reset": "latest",
+        }
+        consumer_config.update(cls._consumer_config())
```

Comment:
Would it make sense to allow the caller to pass in an additional dict to further customize this?

Reply:
Do you think that's still useful if I also switch to the simple constructor that expects the user to pass the consumer?

Reply:
Probably not. 

🔕 _maaaaayyyyybeeeee_ make `consumer_config` a part of the interface rather than hidden so people that want to initialize their own consumer can use it but I don't really feel strongly about that (or even know if it's actually a good idea).